### ISSUE TITLE
[feature request] Improve Include-Page tag to pull only specific content

### ISSUE THREAD

**Issue by PeterMarkuson** [gollum/gollum#1829]
I love Gollum for the include-tag feature (https://github.com/gollum/gollum/wiki#include-tag), because it allows me to assemble and re-use information I have already written down somehwere else. When updating the source of the included information, automatically all other pages using that information are up-to-date immediately 👍 

To be flexible and make full use of that feature, a lot of pages with only small content need to be created and then included in larger pages (in my case setup guides). So, I end up with many atomic pages and little content on each page. Many information blocks build-up on previous ones and would be good to have them on one page together (for editing, review). But, it makes not always sense to include all of it in another page. An improved include-tag with a parameter to select only a block of the page would be a great improvement.

My proposal:

* Today: User needs to create a page for every information he wants to re-use, because the include-tag feature always pulls the full page.
* Improvement: A user could specify a bookmark and in that case only that bookmark and the content below is included in the page.
* Result: Only the content from the bookmark is included. 

The content is identified either from bookmark to the next bookmark or from the level of the heading to the next same level.

**Comment by benjaminwil** [comment]
Thanks for submitting this request.

Based on how `include-tag` is currently implemented, I think what you have proposed would greatly increase the complexity of the existing feature. If we were to implement this, I think it would have to be a a new, separate feature.

Just going to write out my thoughts so far below.

I'm also not sure practically how we'd implement this from a user experience point of view. I think that the proposed feature could be more error prone for writers and editors. Right now the `include-tag` cares very little about the content of the included page. In the proposed feature we now must parse many additional "nodes" and match the right one(s).

I personally see the atomic stub articles that `include-tag` forces a writer to manage as a benefit. While they can "pollute" the wiki file directory, I think that managing them is very straightforward overall, and a team of writers and editors will have no trouble backtracking from an `include-tag` to a specific file with a specific path. I would be curious to hear more about why, in your case, the atomic stub article isn't ideal for you? I'm also curious whether a custom macro could get you closer to what your ideal workflow would be.

**Comment by dometto** [comment]
This is to elaborate a bit on why this would potentially be quite complex, even in a custom macro.

The feature would require the at least partial rendering of the included page. Without a rendered document, we do not have any headings (html `<h1>-<h6>` tags) to include. Rendering the included document *before* it is included is inefficient, but we could theoretically do it. However, when we have the rendered document, there's no easy way to get back to the corresponding line numbers in the source document. That would require the rendered HTML to include a "source map", which not all rendering gems for all markups support. This was the core difficulty behind implementing the [editable sections](https://github.com/gollum/gollum/wiki/5.0-release-notes#editable-sections), and we only solved it on the frontend with JavaScript.

Bottom line is: yes, if there are other ways of improving your workflow, that would be much preferable. :)

**Comment by benjaminwil** [comment]
Thanks for explaining further!

In my own career as a technical writer (which, it's worth mentioning, didn't include my writing very much generic or reusable documentation), much of the shareable pieces of writing ended up _not_ being shared, because it was too complicated to share it across a documentation project in a sophisticated way–even when using my team's preferred documentation writing software MadCap Flare (a piece of software I would recommend no one ever use!!!).

Based on what you've outlined, I think a custom macro for your project could be possible. But I don't know how we could formalize it into a proper feature yet. I would want to survey more people heavily using the `[[include]]` tag to find out what exactly this feature should look like.

That said, here is a [custom macro](https://github.com/gollum/gollum/wiki#custom-macros) that crudely implements what you're describing:

```
module Gollum
  class Macro
    class Snippet < Gollum::Macro
      def render(snippet_name)
        find_snippet(snippet_name)
      end

      private

      def find_snippet(snippet_name)
        snippet_body_regex = /\{{3}\(#{snippet_name}\)(.*)\}{3}/m

        if (match = snippets.match(snippet_body_regex))
          match.captures.first
        else
          nil
        end
      end

      def snippets
        @wiki.pages.detect { |page| page.name == "Snippets" }.text_data
      end
    end
  end
end
```

How does this work? If you have a page with the name "Snippets" (for me: `Snippets.md` in my root directory), you can define snippets in it like this:

```
{{{(snippet one)
Lorem ipsum dolor sit amet, consetetur sadipscing elitr,
sed diam nonumy eirmod tempor invidunt ut labore et dolore
magna aliquyam erat, sed diam voluptua. At vero eos et
accusam et justo duo dolores et ea rebum.
}}}

{{{(snippet two)
Another snippet!
}}}
```

And then, on other pages, call the snippet macro:

```
# Another Page.md

<<Snippet("snippet one")>>
```

On save, it loads the snippet.

**Comment by PeterMarkuson** [comment]
Thank you for your approach. I believe this would be helpful and I tried to implement it. When following your example (adding the two pages, copying the content into it and activating the macro), I get the following error message on "Another page":

**Macro Error for Snippet: undefined method `text_data' for nil:NilClass**

And on Snippets.md I see the `{{{(snippet one)` tag which should be hidden, right?

Also, if I get your example right, the name of the page to source the snippets from is hard-coded into the macro. So it's only looking on "Snippets", but not on any other pages?

**Comment by benjaminwil** [comment]
Yes, this was absolutely a crude example I made in five minutes. And unfortunately my regular expression failed more than I thought it would. 😞 

But it sounds like you’re already tweaking this and making it more suitable for your use case? If you know what Markdown rendering gem your Gollum instance uses (the default is `kramdown`) I believe it’d be possible to build HTML output from your Markdown formatting, too. Check out `kramdown`’s documentation for information on how to do this.

**Comment by PeterMarkuson** [comment]
I started to customize it a bit, but my coding knowledge is very basic and I don't have any experience with Ruby, so it will be hard for me. I added an argument for the pagename to the macro and tried to adopt the regular expression to be more generic, but this is also a challenge for me. 

My current version is:
~~~
module Gollum
  class Macro
    class Snippet2 < Gollum::Macro
      def render(page_name, snippet_name)
        find_snippet(page_name, snippet_name)
      end

      private

      def find_snippet(page_name, snippet_name)
        snippet_body_regex = /#{snippet_name}(.*)#/m

        if (match = snippets(page_name).match(snippet_body_regex))
          match.captures.first
        else
          nil
        end
      end

      def snippets(page_name)
        @wiki.pages.detect { |page| page.name == "#{page_name}" }.text_data
      end
    end
  end
end
~~~

And I call it from another page by `<<Snippet2("Snippets","## Test")>>`

What I would like to achieve:

Calling `<<Snippet2("Snippets","## Test", 2)>>` where "Snippets" is the pagename, "## Test" is the pattern to use as starting point and 2 is the heading-level to stop the include section. 

Example `Snippets` page
~~~
## Test
Here is some text

### Subheading
Additional Text

## Another Heading
Some more text
~~~

This should deliver from the 2nd level heading `## Test` until (but not included) `## Another Heading`.  If the third parameter would be 3 instead of 2, the include should stop before `### Subheading`, because this is a 3rd level heading.

So, the macro needs to analyze the heading level from the starting point. With this mechanism also no special patterns are required in the source text.

Also I changed the markdown engine to github, because with kramdown I had issues with tables and unfortunately have no idea how to convert to html output.

So, thank you for the jumpstart with the Macro which clearly points me to the right direction and would be a suitable solution for me. Unfortunately my skills are not sufficient to develop a full solution on my own. I'll try to educate myself in Ruby, but if your time allows it and it's not too much effort for you, I would appreciate a little more help here.
